Voxelized geometry with material boundaries aligned with voxels

damianjilk2 opened the pull request:
This pull request is an overhaul of the previous DRT script. With updated methods, the purpose is to provide the user with the flux values given certain parameters from input_data.json. 

This PR additionally adds CSV output functionality to the neutron flux calculation. Neutron flux results are now saved to a CSV file named 'results.csv' for easier analysis and visualization. 

The testing file "test_drt_1d.py" has been commented out and will need to be implemented again with the updated methods in a future PR. 

Please provide feedback on the implementation and suggest any improvements or point out any errors that might have been made. Thank you!

damianjilk2 committed:
Initial implementation for voxelized geometry with material boundaries aligned with voxels

damianjilk2 commented on drt_1d.py:
```diff
@@ -1,147 +1,118 @@
+        for j in range(num_voxels):
+            if i == j:
+                # Set diagonal values to small positive value
+                K_trans[i, j] = 1e-20
```
This is an arbitrary value set for the diagonal probabilities. Is there a standard for what this should be?

damianjilk2 replied:
Or perhaps setting the value to zero instead?

gonuke replied:
Given the definition of `K_trans` as returned by `calculate_K_trans`, I think the best answer is 1, since it's the probability that a particle born at the left side of `i` reaches the left side of `i`.

However, this careful interpretation may call into question the utility of the definition of `K_trans` here.

damianjilk2 commented on drt_1d.py:
```diff
@@ -1,147 +1,118 @@
+    min_index = min(ri, rj)
+    max_index = max(ri, rj)
+    distance = abs(ri - rj) * voxel_size
+    tau = np.sum(np.diag(sigma_s_matrix[min_index:max_index])) * voxel_size
```
This calculation takes all cross section values between ri and rj inclusive and sums them. Is this the proper way to find the adjusted cross section when calculating tau?

damianjilk2 replied:
Note: bug was detected with indexing of sigma_s_matrix. Correction is included in PR #18

damianjilk2 commented on test_drt_1d.py:
```diff
@@ -1,45 +1,45 @@
-from drt_1d import (
```
Testing data will be implemented in a future PR

damianjilk2 committed:
Enhance tester program and integrate updates in drt_1d.py for user data validation and error handling.

damianjilk2 committed:
Update CI workflow to trigger on pull requests from any branch

gonuke commented on drt_1d.py:
```diff
@@ -1,147 +1,118 @@
+    phi = np.dot(np.linalg.inv(np.eye(len(K_trans)) - np.dot(K_trans, sigma_s_matrix)), np.dot(K_trans, s))
+    return phi
 
 def main():
```
separation of concerns: it's a good idea to divide up different fundamental tasks into different functions/methods, in this case:
* read & process input
* perform calculation
* write output

This will facilitate testing and allow the calculation step to be reused in slightly different use cases

gonuke commented on drt_1d.py:
```diff
@@ -1,147 +1,118 @@
+    num_voxels = input_data.get("number_of_voxels") 
+    sigma_s_values = input_data.get("scattering_cross_section")
+    sigma_s_matrix = np.diag(sigma_s_values)
+    print("Scattering Cross-section Matrix (sigma_s):\n", sigma_s_matrix)
```
Look into python's `logging` module to provide some more control over output like this.  It's useful now, but become annoying once you are running more cases.

gonuke commented on drt_1d.py:
```diff
@@ -1,147 +1,118 @@
+    min_index = min(ri, rj)
+    max_index = max(ri, rj)
+    distance = abs(ri - rj) * voxel_size
+    tau = np.sum(np.diag(sigma_s_matrix[min_index:max_index])) * voxel_size
```
per our earlier conversation, this should be a good fix for your indexing issue
```suggestion
    tau = np.sum(np.diag(sigma_s_matrix)[min_index:max_index]) * voxel_size
```

gonuke commented on drt_1d.py:
```diff
@@ -1,147 +1,118 @@
+def calculate_K_trans(ri:int, rj:int, voxel_size:float, sigma_s_matrix:np.ndarray):
     """
-    Calculate the total cross section.
+    Calculate the neutron transition probability. Used by the construct_transition_matrix function to build the K_trans matrix.
```
I think this may be best referred to as the streaming operator.  Based on what is returned, this is the probability that a particle born at the left side of `ri` streams to the left side of `rj` without interacting.

gonuke reviewed (changes requested):
A few quick suggestions here.

damianjilk2 committed:
Refactored code with enhanced docstrings and logging improvements

- Improved docstrings for better clarity and consistency across functions.
- Enhanced logging for better traceability and debugging.
- Refined error handling mechanisms, especially in read_input and write_output functions.
- Updated function comments to better reflect their roles and purposes.
- Overall code cleanup for readability and maintainability.

damianjilk2 committed:
Merge branch '1-d-toy-problem-with-voxelized-geometry' into add-tester-script

damianjilk2 committed:
Merge pull request #18 from damianjilk2/add-tester-script

Enhance tester program and integrate updates in drt_1d.py. Merging with PR #14

damianjilk2 commented:
PR #18 has been merged with this PR to allow CI testing to pass. Merge discrepancies have been addressed and resolved. 
Additional changes have been made because of the change request. 
- changed description of K_trans as the streaming operator
- indexing problem of tau (note: update to tau equation regarding issue #22 and slack conversation will be provided in future PR)
-  diagonal streaming operators set to 1
- methods included for reading input data, performing calculations, and writing output
- logging features implemented including extensive logging.DEBUG commands

damianjilk2 commented:
Note: I am not fully confident that this was the preferred way to go about this workflow. 
- commit changes regarding review to PR #14 
- merge PR #14 and PR #18 (18 was not reviewed before merging)
- add comments to PR #14 acknowledging changes and re-request review
I'd love feedback on what the preferred steps would be! Thank you.

gonuke commented:
This workflow was OK in this circumstance. It's true that we would generally prefer to merge one PR before the PR(s) that depend on it.  However, we also generally want all PRs to pass tests, so it's probably more important to make the PR pass tests than to avoid combining these.  Finally, it's best to make PRs as small in scope as possible to facilitate rapid review.

gonuke commented on drt_1d.py:
```diff
@@ -1,147 +1,215 @@
+    if ri < 0 or ri >= sigma_s_matrix.shape[0] or rj < 0 or rj >= sigma_s_matrix.shape[0]:
+        raise ValueError("Voxel indices must be non-negative and within the range of the sigma_s_matrix.")
+    
+    min_index = min(ri, rj)
+    min_index = min(ri, rj)
```
one of these is repeated
```suggestion
    min_index = min(ri, rj)
```

gonuke commented on drt_1d.py:
```diff
@@ -1,147 +1,215 @@
+    min_index = min(ri, rj)
+    max_index = max(ri, rj)
+    distance = abs(ri - rj) * voxel_size
+    tau = np.sum(np.diag(sigma_s_matrix)[min_index:max_index]) * voxel_size
```
This always goes from the left edge to left edge, regardless of which direction the transport is happening because of the way you define min/max.

damianjilk2 replied:
This is a great catch! I will update the description for now to detail the actual calculation. I will solve the issue alongside [#22](https://github.com/damianjilk2/Deterministic-Radiation-Transport-Approximations/issues/22) when computing with two arbitrarily set points. Thank you for the diligent review!

gonuke commented on drt_1d.py:
```diff
@@ -1,147 +1,215 @@
+    num_voxels = input_data.get("number_of_voxels")
+    source_voxel_index = input_data.get("source_voxel_index")
+    sigma_s_values = input_data.get("scattering_cross_section")
+    sigma_s_matrix = np.diag(sigma_s_values)
```
It probably makes more sense to define this as a vector and convert to the matrix in the once place you need it.  Right now, you make the matrix immediately, and tend to extract the vector back out of it in most operations.

gonuke reviewed (changes requested):
Just a few more little comments.  I think this looks pretty good, but as I've mentioned elsewhere, will be hard to verify/validate.

damianjilk2 committed:
Updated .gitignore and removed unnecessary files

damianjilk2 committed:
Refactor sigma_s Handling and Update Copyright Year

- Refactored sigma_s to use a vector for calculations.
- Updated calculate_phi to convert sigma_s to a matrix as needed.
- Simplified code by treating sigma_s as a vector.
- Updated copyright year to 2024.

MicahGale commented on drt_1d.py:
```diff
@@ -1,147 +1,217 @@
+        with open(file_path, "r") as json_file:
+            data = json.load(json_file)
+        logging.debug("Input data: %s", data)
+    except FileNotFoundError:
+        logging.error("File %s not found.", file_path)
+        raise
```
```
except FileNotFoundError as e:

    raise e
```

MicahGale commented on drt_1d.py:
```diff
@@ -1,147 +1,217 @@
+def main():
+    setup_logging()
+    logging.debug("Main function started")
+    input_data = read_input("input_data.json")
```
Implement argparse for input file path.

damianjilk2 commented on drt_1d.py:
```diff
@@ -1,147 +1,217 @@
+def construct_source_term(num_voxels:int, source_voxel_index:int, source_strength:float=1):
     """
-    Calculate the distance between two points along a 1-D geometry.
+    Constructs the source term vector. Assumes a point source. Minor modifications can be made to allow for a distribution.
```
Formatting of docstrings

damianjilk2 commented on drt_1d.py:
```diff
@@ -1,147 +1,217 @@
-# Task 2.2: Construct optical distance along ray
+    logging.debug("Constructing source term with %d voxels, source at index %d, strength %f", num_voxels, source_voxel_index, source_strength)
+    
+    if num_voxels <= 0:
+        raise ValueError("Number of voxels must be a positive integer.")
+    if source_voxel_index < 0 or source_voxel_index >= num_voxels:
+        raise IndexError("Source voxel index must be within the range of voxels.")
```
Additional error handling (e.g. check that values are integers)

damianjilk2 committed:
Refactored code to incorporate feedback: improved logging, argparse handling, and error checking

damianjilk2 commented:
Code has been refactored to account for comments made during Hacker Within meeting. 

Changes include: f-strings for logging; formatting of docstrings; integer error handling; raise e error handling; implementing argparse.

Use for argparse:
`python drt_1d.py --help` shows the user what inputs can be provided and informs them what the default file will be.
`python drt_1d.py --input_file user_file_path` is the command used to run the simulation with `user_file_path`.

gonuke commented on drt_1d.py:
```diff
@@ -1,147 +1,244 @@
+        
+    K_trans = np.zeros((num_voxels, num_voxels))
+    for i in range(num_voxels):
+        for j in range(num_voxels):
+            if i == j:
+                # Set diagonal values to one based on defintion of streaming operator
+                K_trans[i, j] = 1
+            else:
+                K_trans[i, j] = calculate_K_trans(i, j, voxel_size, sigma_s)
```
transition matrix is symmetric, so you can simplify this:
```suggestion
        K_trans[i,i] = 1
        for j in range(i):
            K_trans[i, j] = calculate_K_trans(i, j, voxel_size, sigma_s)
            K_trans[j,i] = K_trans[i,j]
```

damianjilk2 replied:
Great suggestion. This approach simplifies code and optimizes computation time.

gonuke reviewed (changes requested):
This looks pretty good for this degree of approximation - just one comment about simplifying the building of the matrix.

damianjilk2 committed:
Apply transition matrix symmetry

gonuke commented:
Thanks @damianjilk2 - I'm merging this now and we'll move on to better approximations...
